compiler/src/abi/json_builder.rs

```diff
@@ -0,0 +1,240 @@
+use crate::errors::CompileError;
+use serde::Serialize;
+use std::collections::HashMap;
+use vyper_parser::ast as vyp;
+
+/// Used to keep track of custom types defined in a module.
+type TypeDefs<'a> = HashMap<&'a str, &'a vyp::TypeDesc<'a>>;
+
+/// TODO: Add support for events.
+#[derive(Serialize, Debug, PartialEq)]
+pub struct Contract {
+    pub functions: Vec<Function>,
+}
+
+#[derive(Serialize, Debug, PartialEq)]
+pub struct Function {
+    name: String,
+    #[serde(rename = "type")]
+    typ: FunctionType,
+    inputs: Vec<Input>,
+    outputs: Vec<Output>,
+}
+
+#[derive(Serialize, Debug, PartialEq)]
+pub struct Input {
+    name: String,
+    #[serde(rename = "type")]
+    typ: VariableType,
+}
+
+#[derive(Serialize, Debug, PartialEq)]
+pub struct Output {
+    name: String,
+    #[serde(rename = "type")]
+    typ: VariableType,
+}
+
+#[allow(dead_code)]
+#[serde(rename_all = "lowercase")]
+#[derive(Serialize, Debug, PartialEq)]
+pub enum FunctionType {
+    Function,
+    Constructor,
+    Receive,
+    Fallback,
+}
+
+#[serde(rename_all = "lowercase")]
+#[derive(Serialize, Debug, PartialEq)]
+pub enum VariableType {
+    Uint256,
+}
+
+impl ToString for VariableType {
+    fn to_string(&self) -> String {
+        match self {
+            VariableType::Uint256 => String::from("uint256"),
+        }
+    }
+}
+
+#[allow(dead_code)]
+#[serde(rename_all = "lowercase")]
+#[derive(Serialize, Debug, PartialEq)]
+pub enum StateMutability {
+    Pure,
+    View,
+    Nonpayable,
+    Payable,
+}
+
+/// Builds a vector containing contract ABIs.
+pub fn module<'a>(module: &'a vyp::Module) -> Result<Vec<Contract>, CompileError> {
+    let type_defs: TypeDefs<'a> = module
+        .body
+        .iter()
+        .filter_map(|stmt| match stmt.node {
+            vyp::ModuleStmt::TypeDef { ref name, ref typ } => Some((name.node, &typ.node)),
+            _ => None,
+        })
+        .collect();
+
+    module
+        .body
+        .iter()
+        .filter_map(|stmt| match stmt.node {
+            vyp::ModuleStmt::ContractDef { .. } => Some(contract_def(&type_defs, &stmt.node)),
+            _ => None,
+        })
+        .collect::<Result<Vec<Contract>, CompileError>>()
```

Comment:
Yeah, it might be better to have that ~~`ModuleDef`~~ ~~`Module`~~ `ModuleABI` struct and then make the public API:

``` rust
#[derive(Serialize, ...)]
pub struct ModuleABI {
  contracts: HashMap<String, ContractABI>,
}

pub fn build(src: &str) -> Result<ModuleABI, CompileError> { ... }

...

let abi = build(src)?;

// It's still easy to create a json list for a given contract.
let contract_json = abi.get("MyContract").unwrap().serialize();
// or serialize the whole module into a json object
let module_json = abi.serialize();

// or just make use of the module pragmatically 
```